pkg/controller/operators/cleanupcsv_controller.go

```diff
@@ -0,0 +1,151 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/operators/decorators"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/ownerutil"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/metrics"
+)
+
+// CleanupCsvReconciler cleans up cluster-scoped CSV-owned resources.
+type CleanupCsvReconciler struct {
+	client.Client
+
+	log     logr.Logger
+	factory decorators.OperatorFactory
+}
+
+// SetupWithManager adds the operator reconciler to the given controller manager.
+func (r *CleanupCsvReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&operatorsv1alpha1.ClusterServiceVersion{}).
+		Complete(reconcile.Func(r.ReconcileClusterServiceVersion))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewCleanupCsvReconciler constructs and returns an CleanupCsvReconciler.
+// As a side effect, the given scheme has operator discovery types added to it.
+func NewCleanupCsvReconciler(cli client.Client, log logr.Logger, scheme *runtime.Scheme) (*CleanupCsvReconciler, error) {
+	// Add watched types to scheme.
+	if err := AddToScheme(scheme); err != nil {
+		return nil, err
+	}
+
+	factory, err := decorators.NewSchemedOperatorFactory(scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CleanupCsvReconciler{
+		Client: cli,
+
+		log:     log,
+		factory: factory,
+	}, nil
+}
+
+// ReconcileClusterServiceVersion projects the component labels of a given CSV onto all resources owned by it.
+func (r *CleanupCsvReconciler) ReconcileClusterServiceVersion(ctx context.Context, req ctrl.Request) (reconcile.Result, error) {
+	// Set up a convenient log object so we don't have to type request over and over again
+	log := r.log.WithValues("request", req)
+	log.V(1).Info("reconciling csv")
+	metrics.EmitCleanupCSVReconcile(req.Namespace, req.Name)
+
+	// Fetch the CSV from the cache
+	in := &operatorsv1alpha1.ClusterServiceVersion{}
+	if err := r.Get(ctx, req.NamespacedName, in); err != nil {
+		if apierrors.IsNotFound(err) {
+			err = nil
+		} else {
+			log.Error(err, "Error finding ClusterServiceVersion")
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Ignore copied csvs
+	if in.IsCopied() {
+		return reconcile.Result{}, nil
+	}
+
+	myFinalizerName := "olm.csv.cleanup.controller/finalizer"
+	if in.ObjectMeta.DeletionTimestamp.IsZero() {
+		// not being deleted, add finalizer if not present
+		if !controllerutil.ContainsFinalizer(in, myFinalizerName) {
+			controllerutil.AddFinalizer(in, myFinalizerName)
+			if err := r.Update(ctx, in); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
```

Comment:
Same with C, but in this case (which I'd have to look at closer), it may not be a `return` (and doesn't appear to be from the code snippet above) for the whole `if` statement, meaning that the `else` block shouldn't necessarily be executed. (There's really not enough context here without diving back into the code, and it's Sunday, and I'm cleaning up after a snowstorm.)